apm-agent-core/src/main/java/co/elastic/apm/agent/collections/LongList.java

```diff
@@ -58,9 +59,20 @@ public void addAll(LongList other) {
         size += other.size;
     }
 
-    private void ensureCapacity(int size) {
-        if (longs.length < size) {
-            longs = Arrays.copyOf(longs, longs.length * 2);
+    private void ensureCapacity(long minCapacity) {
+        if (longs.length < minCapacity) {
+            longs = Arrays.copyOf(longs, newCapacity(minCapacity, longs.length));
+        }
+    }
+
+    static int newCapacity(long minCapacity, long oldCapacity) {
+        long growBy50Percent = oldCapacity + (oldCapacity >> 1);
+        if (minCapacity <= growBy50Percent) {
+            return (int) growBy50Percent;
+        } else if (minCapacity <= MAX_ARRAY_SIZE) {
+            return (int) minCapacity;
+        } else {
+            throw new OutOfMemoryError();
```

Comment:
That's exactly how it's done in ArrayList, for example.

> If you are OK with OutOfMemoryError being thrown here, why even check that the max is exceeded?

Because it would not throw an `OutOfMemoryError` otherwise, the int would just overflow and be negative.

Reply:
That's true. It's weird that ArrayList does that, this error indicates failure to allocate memory.
Bottom line, I think this is extremely high and we should not consume so much heap memory if something goes bad. We need the feature not to work if we have a bug, but not the JVM getting crashed or sluggish. 
It's a data collection, so you can add a constructor to limit capacity, but since it is only used by us, I think limiting the capacity internally to MUCH lower is good enough.

Reply:
Sounds a bit like a slippery slope: following that argumentation, we'd have to limit the size of all the collections used in the agent. What makes the usage of this list special or more risky?

Reply:
It is a slippery slope, just overthought this because of the specific issue with this one.